test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -119,119 +126,98 @@ func BenchmarkPerfScheduling(b *testing.B) {
 		klog.Fatalf("%v", err)
 	}
 	defer nodePreparer.CleanupNodes()
 
-	config := testutils.NewTestPodCreatorConfig()
-	config.AddStrategy(setupNamespace, test.InitPods.Num, setupPodStrategy)
-	podCreator := testutils.NewTestPodCreator(clientset, config)
-	podCreator.CreatePods()
-
-	for {
-		scheduled, err := getScheduledPods(podInformer)
-		if err != nil {
-			klog.Fatalf("%v", err)
-		}
-		if len(scheduled) >= test.InitPods.Num {
-			break
-		}
-		klog.Infof("got %d existing pods, required: %d", len(scheduled), test.InitPods.Num)
-		time.Sleep(1 * time.Second)
-	}
-
-	scheduled := int32(0)
-	completedCh := make(chan struct{})
-	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		UpdateFunc: func(old, cur interface{}) {
-			curPod := cur.(*v1.Pod)
-			oldPod := old.(*v1.Pod)
-
-			if len(oldPod.Spec.NodeName) == 0 && len(curPod.Spec.NodeName) > 0 {
-				if atomic.AddInt32(&scheduled, 1) >= int32(test.PodsToSchedule.Num) {
-					completedCh <- struct{}{}
-				}
-			}
-		},
-	})
+	createPods(setupNamespace, test.InitPods, clientset)
+	waitNumPodsScheduled(test.InitPods.Num, podInformer)
 
 	// start benchmark
 	b.ResetTimer()
 
-	// Start measuring throughput
+	// Start test data collectors.
 	stopCh := make(chan struct{})
-	throughputCollector := newThroughputCollector(podInformer)
-	go throughputCollector.run(stopCh)
+	collectors := getTestDataCollectors(test, podInformer)
+	for _, collector := range collectors {
+		go collector.run(stopCh)
+	}
 
-	// Scheduling the main workload
-	config = testutils.NewTestPodCreatorConfig()
-	config.AddStrategy(testNamespace, test.PodsToSchedule.Num, testPodStrategy)
-	podCreator = testutils.NewTestPodCreator(clientset, config)
-	podCreator.CreatePods()
+	// Schedule the main workload
+	createPods(testNamespace, test.PodsToSchedule, clientset)
+	waitNumPodsScheduled(test.InitPods.Num+test.PodsToSchedule.Num, podInformer)
 
-	<-completedCh
 	close(stopCh)
-
 	// Note: without this line we're taking the overhead of defer() into account.
 	b.StopTimer()
 
-	setNameLabel := func(dataItem *DataItem) DataItem {
-		if dataItem.Labels == nil {
-			dataItem.Labels = map[string]string{}
+	var dataItems []DataItem
+	for _, collector := range collectors {
+		dataItems = append(dataItems, collector.collect(b)...)
+	}
+	return dataItems
+}
+
+func waitNumPodsScheduled(num int, podInformer coreinformers.PodInformer) {
+	for {
+		scheduled, err := getScheduledPods(podInformer)
+		if err != nil {
+			klog.Fatalf("%v", err)
 		}
-		dataItem.Labels["Name"] = b.Name()
-		return *dataItem
+		if len(scheduled) >= num {
+			break
+		}
+		klog.Infof("got %d existing pods, required: %d", len(scheduled), num)
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func getTestDataCollectors(tc testCase, podInformer coreinformers.PodInformer) []testDataCollector {
+	collectors := []testDataCollector{newThroughputCollector(podInformer)}
+	metricsCollector := tc.MetricsCollector
+	if metricsCollector == nil {
```

Comment:
I don't see `tc.MetricsCollector` ever being set in this PR. Do you plan to implement it in a follow-up PR? Although, I would expect to have `tc.MetricsCollector` as `[]string` (and renamed to `tc.Metrics` or similar)  and run `newMetricsCollector` with it in case it's not nil:
```go
if metricsCollector == nil {
	metricsCollector = newMetricsCollector(defaultMetrics)
} else {
	metricsCollector = newMetricsCollector(tc.Metrics)
}
```

Reply:
I created a `metricCollectorConfig` that contains the list of metrics. The idea is that each collector can have its own config that can be mapped to the test case config in YAML

Reply:
Yeah, I don't think it's a good practice to do it this way. I rather have an explicit initialization through a single point of failure instead of populating the struct directly. The struct or the initialization can change in the future and it can be easy to forget there's not just `newMetricsCollector`.